### ISSUE TITLE
Unable to resume cached task when process tag accesses object fields or methods.

### ISSUE THREAD

**Issue by robsyme** [nextflow-io/nextflow#1811]
### Expected Behaviour / Actual Behavior

**Expected:** I would expect that setting the `tag` directive would not affect the ability for Nextflow to re-use cached results.
**Actual:** If the input channel contains a Groovy object and the `tag` directive accesses that object's fields or methods, Nextflow is unable to re-use cached results.

### Steps to Reproduce the Problem

I've got a minimal workflow available at [robsyme/nextflow-bug-report](https://github.com/robsyme/nextflow-bug-report) that replicates the behaviour. You can reproduce the bug with:

```
nextflow run -r main robsyme/nextflow-bug-report
nextflow run -r main -resume robsyme/nextflow-bug-report
```

### Program Output 

Example runs (once to set the cache, and once to attempt cache re-use). Nextflow warns that the cache could not be used, and suggests that the log includes details, but there are no extra details in the log.

```
❯ rm -rf .nextflow* work

❯ nextflow run -r main robsyme/nextflow-bug-report
N E X T F L O W  ~  version 20.04.1
Launching `robsyme/nextflow-bug-report` [kickass_mayer] - revision: 5a7f833e04 [main]
WARN: DSL 2 IS AN EXPERIMENTAL FEATURE UNDER DEVELOPMENT -- SYNTAX MAY CHANGE IN FUTURE RELEASE
executor >  local (9)
[b3/442385] process > NoTag (2)             [100%] 3 of 3 ✔
[cf/02094d] process > TagUsesField (Lassie) [100%] 3 of 3 ✔
[28/62159a] process > TagUsesMethod (Dog)   [100%] 3 of 3 ✔

❯ nextflow run -r main -resume robsyme/nextflow-bug-report
N E X T F L O W  ~  version 20.04.1
Launching `robsyme/nextflow-bug-report` [angry_stallman] - revision: 5a7f833e04 [main]
WARN: DSL 2 IS AN EXPERIMENTAL FEATURE UNDER DEVELOPMENT -- SYNTAX MAY CHANGE IN FUTURE RELEASE
executor >  local (6)
[72/26d6b2] process > NoTag (3)           [100%] 3 of 3, cached: 3 ✔
[1e/6b961e] process > TagUsesField (Spot) [100%] 3 of 3 ✔
[50/d66c73] process > TagUsesMethod (Dog) [100%] 3 of 3 ✔
WARN: [TagUsesMethod (Dog)] Unable to resume cached task -- See log file for details
WARN: [TagUsesField (Lassie)] Unable to resume cached task -- See log file for details
WARN: [TagUsesField (Fido)] Unable to resume cached task -- See log file for details
WARN: [TagUsesField (Spot)] Unable to resume cached task -- See log file for details
```

### Example Workflow Details

We have a Groovy "Dog" class defined in [`lib/Dog.groovy`](https://github.com/robsyme/nextflow-bug-report/blob/main/lib/Dog.groovy). I override the `hashCode` method so that the dog's hash is consistent between workflow invocations. Without the `hashCode()` override, two dog objects with the same name will produce different hashes and any process with the Dog as input will **always** re-run and nextflow will never check for the cached copy.

```
class Dog {
    String name

    String toString() { "Dog" }

    @Override
    public int hashCode() { name.hashCode() }
}
```

In [`main.nf`](https://github.com/robsyme/nextflow-bug-report/blob/main/main.nf), we make a channel that is loaded with four 'dogs' which are passed through three processes. 

As seen in the shell output above - if the process has a `tag` directive that accesses the object's field or calls a method on the object (processes `TagUsesField` and `TagUsesMethod`), Nextflow is unable to use any cached results. Exactly the same process without a `tag` directive (process `NoTag`) is able to use the cached copy without issue.

```
Channel.from("Fido", "Spot", "Lassie") | map { new Dog(name: it) } | (NoTag & TagUsesField & TagUsesMethod)
```

The three process are:

```
process NoTag {
    input:
    val(dog)

    "touch test.txt"
}

process TagUsesField {
    tag { dog.name }
    input:
    val(dog)

    "touch test.txt"
}

process TagUsesMethod {
    tag { dog }
    input:
    val(dog)

    "touch test.txt"
}
```

### Additional Details

Log file (`.nextflow.log`) from the second run:

```
Nov-26 18:18:03.158 [main] DEBUG nextflow.cli.Launcher - Setting http proxy: [192.168.32.1, 3128]
Nov-26 18:18:03.210 [main] DEBUG nextflow.cli.Launcher - Setting https proxy: [192.168.32.1, 3128]
Nov-26 18:18:03.211 [main] DEBUG nextflow.cli.Launcher - $> nextflow run -r main -resume robsyme/nextflow-bug-report
Nov-26 18:18:03.272 [main] INFO  nextflow.cli.CmdRun - N E X T F L O W  ~  version 20.10.0
Nov-26 18:18:03.310 [main] DEBUG nextflow.scm.AssetManager - Project manifest does not exist: /home/bravolims-dev/.nextflow/assets/robsyme/nextflow-bug-report/nextflow.config (No such file or directory)
Nov-26 18:18:03.317 [main] DEBUG nextflow.scm.AssetManager - Git config: /home/bravolims-dev/.nextflow/assets/robsyme/nextflow-bug-report/.git/config; branch: master; remote: origin; url: https://github.com/robsyme/nextflow-bug-report.git
Nov-26 18:18:03.331 [main] DEBUG nextflow.scm.AssetManager - Git config: /home/bravolims-dev/.nextflow/assets/robsyme/nextflow-bug-report/.git/config; branch: master; remote: origin; url: https://github.com/robsyme/nextflow-bug-report.git
Nov-26 18:18:03.587 [main] INFO  nextflow.cli.CmdRun - Launching `robsyme/nextflow-bug-report` [hungry_wright] - revision: 668ceafce4 [main]
Nov-26 18:18:04.357 [main] DEBUG nextflow.Session - Session uuid: 96d15307-4521-4ff4-a5a7-9f76b5107276
Nov-26 18:18:04.357 [main] DEBUG nextflow.Session - Run name: hungry_wright
Nov-26 18:18:04.357 [main] DEBUG nextflow.Session - Executor pool size: 48
Nov-26 18:18:04.375 [main] DEBUG nextflow.cli.CmdRun -
  Version: 20.10.0 build 5430
  Created: 01-11-2020 15:14 UTC (10:14 EDT)
  System: Linux 3.10.0-957.el7.x86_64
  Runtime: Groovy 3.0.5 on OpenJDK 64-Bit Server VM 1.8.0_191-b12
  Encoding: UTF-8 (UTF-8)
  Process: 2819791@abacus3 [10.0.1.22]
  CPUs: 48 - Mem: 125.4 GB (81.7 GB) - Swap: 12 GB (12 GB)
Nov-26 18:18:04.398 [main] DEBUG nextflow.Session - Work-dir: /tmp/work [xfs]
Nov-26 18:18:04.399 [main] DEBUG nextflow.Session - Script base path does not exist or is not a directory: /home/bravolims-dev/.nextflow/assets/robsyme/nextflow-bug-report/bin
Nov-26 18:18:04.462 [main] DEBUG nextflow.Session - Observer factory: TowerFactory
Nov-26 18:18:04.464 [main] DEBUG nextflow.Session - Observer factory: DefaultObserverFactory
Nov-26 18:18:04.665 [main] DEBUG nextflow.Session - Session start invoked
Nov-26 18:18:04.683 [main] DEBUG nextflow.Session - Using default localLib path: /home/bravolims-dev/.nextflow/assets/robsyme/nextflow-bug-report/lib
Nov-26 18:18:04.690 [main] DEBUG nextflow.Session - Adding to the classpath library: /home/bravolims-dev/.nextflow/assets/robsyme/nextflow-bug-report/lib
Nov-26 18:18:04.760 [main] WARN  nextflow.NextflowMeta$Preview - DSL 2 IS AN EXPERIMENTAL FEATURE UNDER DEVELOPMENT -- SYNTAX MAY CHANGE IN FUTURE RELEASE
Nov-26 18:18:05.546 [main] DEBUG nextflow.script.ScriptRunner - > Launching execution
Nov-26 18:18:05.665 [main] DEBUG nextflow.Session - Workflow process names [dsl2]: NoTag, TagUsesMethod, TagUsesField
Nov-26 18:18:05.782 [main] DEBUG nextflow.executor.ExecutorFactory - << taskConfig executor: null
Nov-26 18:18:05.782 [main] DEBUG nextflow.executor.ExecutorFactory - >> processorType: 'local'
Nov-26 18:18:05.786 [main] DEBUG nextflow.executor.Executor - [warm up] executor > local
Nov-26 18:18:05.795 [main] DEBUG n.processor.LocalPollingMonitor - Creating local task monitor for executor 'local' > cpus=48; memory=125.4 GB; capacity=48; pollInterval=100ms; dumpInterval=5m
Nov-26 18:18:05.912 [main] DEBUG nextflow.executor.ExecutorFactory - << taskConfig executor: null
Nov-26 18:18:05.912 [main] DEBUG nextflow.executor.ExecutorFactory - >> processorType: 'local'
Nov-26 18:18:05.916 [main] DEBUG nextflow.executor.ExecutorFactory - << taskConfig executor: null
Nov-26 18:18:05.916 [main] DEBUG nextflow.executor.ExecutorFactory - >> processorType: 'local'
Nov-26 18:18:05.920 [main] DEBUG nextflow.Session - Ignite dataflow network (4)
Nov-26 18:18:05.921 [main] DEBUG nextflow.processor.TaskProcessor - Starting process > NoTag
Nov-26 18:18:05.921 [main] DEBUG nextflow.processor.TaskProcessor - Starting process > TagUsesField
Nov-26 18:18:05.921 [main] DEBUG nextflow.processor.TaskProcessor - Starting process > TagUsesMethod
Nov-26 18:18:05.922 [main] DEBUG nextflow.script.ScriptRunner - > Await termination
Nov-26 18:18:05.922 [main] DEBUG nextflow.Session - Session await
Nov-26 18:18:06.010 [Actor Thread 10] DEBUG nextflow.util.CacheHelper - [WARN] Unknown hashing type: class Dog
Nov-26 18:18:06.128 [Actor Thread 13] WARN  nextflow.processor.TaskProcessor - [TagUsesMethod (Dog)] Unable to resume cached task -- See log file for details
Nov-26 18:18:06.131 [Actor Thread 9] WARN  nextflow.processor.TaskProcessor - [TagUsesField (Fido)] Unable to resume cached task -- See log file for details
Nov-26 18:18:06.131 [Actor Thread 14] WARN  nextflow.processor.TaskProcessor - [TagUsesField (Lassie)] Unable to resume cached task -- See log file for details
Nov-26 18:18:06.131 [Actor Thread 12] WARN  nextflow.processor.TaskProcessor - [TagUsesField (Spot)] Unable to resume cached task -- See log file for details
Nov-26 18:18:06.141 [Actor Thread 8] INFO  nextflow.processor.TaskProcessor - [c3/6ebbcd] Cached process > NoTag (1)
Nov-26 18:18:06.141 [Actor Thread 15] INFO  nextflow.processor.TaskProcessor - [3a/724b2a] Cached process > NoTag (3)
Nov-26 18:18:06.141 [Actor Thread 11] INFO  nextflow.processor.TaskProcessor - [3d/4c9fd5] Cached process > NoTag (2)
Nov-26 18:18:06.185 [Task submitter] DEBUG nextflow.executor.LocalTaskHandler - Launch cmd line: /bin/bash -ue .command.run
Nov-26 18:18:06.192 [Task submitter] INFO  nextflow.Session - [86/c6b1ad] Submitted process > TagUsesMethod (Dog)
Nov-26 18:18:06.198 [Task submitter] DEBUG nextflow.executor.LocalTaskHandler - Launch cmd line: /bin/bash -ue .command.run
Nov-26 18:18:06.198 [Task submitter] INFO  nextflow.Session - [e9/7461e3] Submitted process > TagUsesMethod (Dog)
Nov-26 18:18:06.207 [Task submitter] DEBUG nextflow.executor.LocalTaskHandler - Launch cmd line: /bin/bash -ue .command.run
Nov-26 18:18:06.208 [Task submitter] INFO  nextflow.Session - [3b/e4fa6f] Submitted process > TagUsesField (Lassie)
Nov-26 18:18:06.212 [Task submitter] DEBUG nextflow.executor.LocalTaskHandler - Launch cmd line: /bin/bash -ue .command.run
Nov-26 18:18:06.213 [Task submitter] INFO  nextflow.Session - [06/591370] Submitted process > TagUsesMethod (Dog)
Nov-26 18:18:06.217 [Task submitter] DEBUG nextflow.executor.LocalTaskHandler - Launch cmd line: /bin/bash -ue .command.run
Nov-26 18:18:06.218 [Task submitter] INFO  nextflow.Session - [e5/097308] Submitted process > TagUsesField (Fido)
Nov-26 18:18:06.220 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[id: 3; name: TagUsesMethod (Dog); status: COMPLETED; exit: 0; error: -; workDir: /tmp/work/86/c6b1ad0003d2d63e7ce397f05b0ef4]
Nov-26 18:18:06.222 [Task submitter] DEBUG nextflow.executor.LocalTaskHandler - Launch cmd line: /bin/bash -ue .command.run
Nov-26 18:18:06.222 [Task submitter] INFO  nextflow.Session - [9f/116cae] Submitted process > TagUsesField (Spot)
Nov-26 18:18:06.225 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[id: 7; name: TagUsesMethod (Dog); status: COMPLETED; exit: 0; error: -; workDir: /tmp/work/e9/7461e3edac392349d10c51e784f857]
Nov-26 18:18:06.226 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[id: 2; name: TagUsesField (Lassie); status: COMPLETED; exit: 0; error: -; workDir: /tmp/work/3b/e4fa6f364ca1928f04d4f528d42721]
Nov-26 18:18:06.228 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[id: 9; name: TagUsesMethod (Dog); status: COMPLETED; exit: 0; error: -; workDir: /tmp/work/06/591370e3684a67437eec6c460865cd]
Nov-26 18:18:06.237 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[id: 8; name: TagUsesField (Fido); status: COMPLETED; exit: 0; error: -; workDir: /tmp/work/e5/0973085b771b73eb0c30c705c7a39f]
Nov-26 18:18:06.238 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[id: 1; name: TagUsesField (Spot); status: COMPLETED; exit: 0; error: -; workDir: /tmp/work/9f/116caebf1d77e0b08814f288b44510]
Nov-26 18:18:06.239 [main] DEBUG nextflow.Session - Session await > all process finished
Nov-26 18:18:06.239 [main] DEBUG nextflow.Session - Session await > all barriers passed
Nov-26 18:18:06.248 [main] DEBUG nextflow.trace.WorkflowStatsObserver - Workflow completed > WorkflowStats[succeededCount=6; failedCount=0; ignoredCount=0; cachedCount=3; pendingCount=0; submittedCount=0; runningCount=0; retriesCount=0; abortedCount=0; succeedDuration=36ms; failedDuration=0ms; cachedDuration=3ms;load
Nov-26 18:18:06.333 [main] DEBUG nextflow.CacheDB - Closing CacheDB done
Nov-26 18:18:06.341 [main] DEBUG nextflow.script.ScriptRunner - > Execution complete -- Goodbye
```

### Environment 

Confirmed on macOS:

* Nextflow version: 20.04.1
* Java version: openjdk version "1.8.0_212"
* Operating system: macOS
* Bash version: zsh 5.8

Confirmed on Linux:

* Nextflow version: 20.10.0
* Java version: openjdk version "1.8.0_191"
* Operating system: Linux
* Bash version: GNU bash, version 4.2.46(2)-release (x86_64-redhat-linux-gnu)

**Comment by robsyme** [comment]
After a suggestion from Paolo, I tested that the behaviour persisted even if the class implemented both Serializable and [CacheFunnel](https://github.com/nextflow-io/nextflow/blob/a005b1b0c6cdaced5ce0a781498f384aaaad8815/modules/nf-commons/src/main/nextflow/util/CacheFunnel.groovy#L12-L16). I've [updated the example repository](https://github.com/robsyme/nextflow-bug-report/commit/7abe67daed35dba092e5094c8cfafdc295b27837) to reflect the attempt.

**Comment by robsyme** [comment]
An exception is raised during the `[getTaskEntry](https://github.com/nextflow-io/nextflow/blob/master/modules/nextflow/src/main/groovy/nextflow/CacheDB.groovy#L155:15)` function. When accessing the object's property in the `tag` directive.

The Exception is:

```
com.esotericsoftware.kryo.KryoException: Unable to find class: Dog
```

In the Kyro docs in the [Registration section](https://github.com/EsotericSoftware/kryo/blob/master/README.md#registration), it mentions that 

> all classes that Kryo will read or write must be registered beforehand

Registering a class is done with the `kryo.register(SomeClass.class)`.  Nextflow already registers a number of classes when setting up the Kryo object in [SerializationHelper.groovy#L84-L115](https://github.com/nextflow-io/nextflow/blob/553c3119cc81012c3c4667ba280f174b0a7c7f8b/modules/nextflow/src/main/groovy/nextflow/util/SerializationHelper.groovy#L84-L115) 

Is it possible to get user-defined classes also added to the list of classes registered by Kryo?

**Comment by pditommaso** [comment]
Yes, this is the problem. Unfortunately, there isnt' a built-in mechanism to register custom serialisers. What trying to stringify your custom object ie. 

```
    tag "${dog.toString()}"
```

instead of 

```
    tag { dog }
```

In principle also `tag "$dog"` should work.

**Comment by robsyme** [comment]
Ah, of course! Thanks Paolo. I appreciate the pointer.

**Comment by pditommaso** [comment]
Let me know if it helps

**Comment by robsyme** [comment]
Ah, regrettably switching the `tag` directive to a string rather than a closure did not fix the problem. What really surprised me was that even setting the tag to a static string that doesn't interact with the object still prevents the cache being used:

```
process StaticTag {
    tag "$custom_tag"

    input:
    tuple custom_tag, dog

    "touch test.txt"
}

workflow {
    dogs = Channel.from("Fido", "Spot", "Lassie") | map { new Dog(name: it) }
    dogs | map { ["FOOBAR", it] } | StaticTag
}
```

I've updated the example repo (branch `static_tag`) to use this more minimal example:

```
❯ rm -rf .nextflow*; nextflow run -r static_tag robsyme/nextflow-bug-report && nextflow run -r static_tag robsyme/nextflow-bug-report -resume
N E X T F L O W  ~  version 20.11.0-edge
Launching `robsyme/nextflow-bug-report` [loving_marconi] - revision: 199b6ce2e9 [static_tag]
WARN: DSL 2 IS AN EXPERIMENTAL FEATURE UNDER DEVELOPMENT -- SYNTAX MAY CHANGE IN FUTURE RELEASE
executor >  local (3)
[6d/3ccce5] process > StaticTag (FOOBAR) [100%] 3 of 3 ✔

N E X T F L O W  ~  version 20.11.0-edge
Launching `robsyme/nextflow-bug-report` [zen_swanson] - revision: 199b6ce2e9 [static_tag]
WARN: DSL 2 IS AN EXPERIMENTAL FEATURE UNDER DEVELOPMENT -- SYNTAX MAY CHANGE IN FUTURE RELEASE
executor >  local (3)
[85/20fef2] process > StaticTag (FOOBAR) [100%] 3 of 3 ✔
WARN: [StaticTag (FOOBAR)] Unable to resume cached task -- See log file for details```

**Comment by pditommaso** [comment]
That's a tricky matter. The long story is that the task evaluation context is serialised to that it can be re-evaluated during the resume. For some reason the `dog` enters in the list of serialised objects 

https://github.com/nextflow-io/nextflow/blob/6c112b8fe25bc6e92205af63cb20a777dc347027/modules/nextflow/src/main/groovy/nextflow/processor/TaskContext.groovy#L219-L224

**Comment by robsyme** [comment]
Great. Thanks again for the pointer. 

As I understand it, our dog ends up in the task context in both processes the process with the tag and the process without the tag. In both cases, the dog is added during [`makeTaskContextStage1`](https://github.com/nextflow-io/nextflow/blob/9518706cf42b844e08ce1763ee585a9434279be4/modules/nextflow/src/main/groovy/nextflow/processor/TaskProcessor.groovy#L1856). The cache works for the `NoTag` process but not the `StaticTag` process because the task context is not serialized for the `NoTag` process.

After looking through the SerializationHelper, TaskContext and TaskProcessor code, I don't think it's going to be easy to 

I think that if the presence of a tag directive forces serialization of the full task context, it's going to be difficult to support resumption of processes that consume custom Groovy classes.

**Comment by robsyme** [comment]
I've also tried implementing KryoSerializable on the class, but this doesn't help either. I can confirm that implementing `CacheFunnel` is required for the caching to work on the `NoTag` process, but `CacheFunnel` does not fix the cache rejection in processes that have a tag set.

```groovy
class Dog implements CacheFunnel, KryoSerializable {
    String name

    String toString() {
        "Dog: $name"
    }

    void write (Kryo kryo, Output output) {
        output.writeString(name);
    }

    void read (Kryo kryo, Input input) {
        name = kryo.readObject(input, String.class);
    }

    Hasher funnel(Hasher hasher, HashMode mode) {
        return CacheHelper.hasher(hasher, this.name, mode)
    }
}
```

**Comment by robsyme** [comment]
I've also tried building Nextflow with `kryo.setRegistrationRequired(false)` in SerializationHelper.groovy, but this doesn't seem to help either.

**Comment by pditommaso** [comment]
Sorry for the late reply Rob. Closing this in favour of #1934

**Comment by robsyme** [comment]
Not a problem! I enjoyed getting an opportunity to dig around a bit, even if I didn't find a clean solution. I like the look of the other fix, thanks for the pointer.

**Comment by pditommaso** [comment]
See https://github.com/nextflow-io/nextflow/issues/1934#issuecomment-791506523

**Comment by sstadick** [comment]
@robsyme, does running on `21.04.0-edge` fix the tag issue for you? I'm still seeing `com.esotericsoftware.kryo.KryoException: Unable to find class:` when accessing a `ValueObject` in the tag.

**Comment by sstadick** [comment]
@pditommaso , pulling the example repo and running this under `21.04.0-edge`, this is not fixed by the `ValueObject` introduced [here](https://github.com/nextflow-io/nextflow/issues/1934). Using an object in a `tag` seems to still result in it not being cacheable. I can open a fresh issue if that would be preferable. 

Here's my [PR](https://github.com/robsyme/nextflow-bug-report/pull/1) to run @robsyme's excellent example repo under `21.04.0-edge`

**Comment by sstadick** [comment]
But from the comments above, it seems like using an object in a tag may just not work with cache? Which is okay as well!

**Comment by sstadick** [comment]
The following fixes it under `21.04.0-edge`

```
import nextflow.io.ValueObject
import nextflow.util.KryoHelper

@ValueObject
class Dog  {

    static {
        KryoHelper.register(Dog)
    }

    String name
}
```
